File: app/src/pushEnv.js
// TELEFON BİLDİRİMİ — ORTAM VE YETENEK SINIFLANDIRMASI (SAF MODÜL)
//
// React Native bağımlılığı YOKTUR (pushPlanner.js ile aynı gerekçe): node
// testlerinde doğrudan çalıştırılabilsin diye. Burada `Platform` import
// EDİLMEZ; platform adı dışarıdan verilir.
//
// NEDEN AYRI BİR DOSYA — GERÇEK HATA BURADAN ÇIKTI:
//   Eski kod "destekleniyor mu?" sorusunu tek satırda yanıtlıyordu:
//       return !isWeb && !!Notifications;
//   ve `require('expo-notifications')` hatası sessizce yutuluyordu. Gerçek
//   Android telefonda paket kurulu olmasına rağmen bu import PATLAYABİLİYOR
//   (expo-notifications'ın barrel dosyası, YEREL bildirimle ilgisi olmayan
//   uzak-push modüllerini de modül gövdesinde yüklüyor; Expo Go içinde bu
//   yerli modüller kayıtlı değil ve `requireNativeModule` hata fırlatıyor).
//   Sonuç: gerçek telefon "web" gibi görünüyor ve ekranda "Tarayıcıda telefon
//   bildirimi kurulamaz." yazıyordu. YANLIŞ TEŞHİS.
//
// KESİN KURALLAR:
//  1) Karar YALNIZ iki gerçeğe dayanır: (a) gerçek platform, (b) yerel bildirim
//     API'sinin elde olup olmadığı. Expo Go, `appOwnership`, geliştirme modu ya
//     da config plugin durumu TEK BAŞINA "desteklenmiyor" gerekçesi DEĞİLDİR ve
//     bu dosyaya girdi olarak bile alınmaz.
//  2) Modül yükleme hatası sessizce "tarayıcı" mesajına dönüşmez; gerçek
//     teknik durum ayrıştırılır ve hata metni saklanır.
//  3) Çalışan bir modül, önceki bir yükleme hatasını geçersiz kılar: barrel
//     patlayıp parçalı yükleme başardıysa durum HAZIR'dır (hata yalnız uyarı
//     olarak taşınır) — çünkü kullanıcı için önemli olan bildirimin kurulması.
//  4) Uydurma yok: "hazır" demek için gereken işlevlerin gerçekten fonksiyon
//     olduğu tek tek doğrulanır.

/** Ortamın ayrıştırılmış hâlleri. */
export const DURUM = {
  WEB: 'web',                 // gerçekten tarayıcı → yerel bildirim yok
  HAZIR: 'hazir',             // yerel bildirim API'si elde, kurulabilir
  MODUL_YOK: 'modul-yok',     // paket bulunamadı (kurulu değil)
  MODUL_HATA: 'modul-hata',   // paket var ama yüklenirken hata verdi
  API_EKSIK: 'api-eksik',     // modül yüklendi ama gereken işlevler yok
};

/** Yerel bildirim için ZORUNLU işlevler — biri yoksa "hazır" denemez. */
export const GEREKLI_API = [
  'getPermissionsAsync',
  'requestPermissionsAsync',
  'scheduleNotificationAsync',
  'cancelScheduledNotificationAsync',
  'getAllScheduledNotificationsAsync',
];

/**
 * Olsa iyi olur ama olmadan da bildirim kurulabilir:
 *  - setNotificationChannelAsync : yalnız Android kanalı (yoksa sistem
 *    varsayılan kanalı kullanılır)
 *  - setNotificationHandler      : uygulama açıkken banner davranışı
 *  - addNotificationResponseReceivedListener : bildirime dokununca yönlendirme
 */
export const SECMELI_API = [
  'setNotificationChannelAsync',
  'setNotificationHandler',
  'addNotificationResponseReceivedListener',
];

function hataMetni(e) {
  if (!e) return '';
  if (typeof e === 'string') return e.trim();
  const m = e?.message != null ? String(e.message) : String(e);
  return m.trim();
}

function eksikOlanlar(modul, liste) {
  const eksik = [];
  for (const ad of liste) {
    if (typeof modul?.[ad] !== 'function') eksik.push(ad);
  }
  return eksik;
}

/**
 * Ortamı sınıflandırır.
 *
 * @param {object} g
 * @param {string} [g.platformOS]     'web' | 'android' | 'ios' | ...
 * @param {object} [g.modul]          yüklenebilmiş bildirim API'si (ya da null)
 * @param {Error|string} [g.yuklemeHatasi] modül yüklenemediyse GERÇEK hata
 * @param {string} [g.kaynak]         'paket' (tek parça) | 'parcali' (yedek yol)
 * @returns {{durum:string, destek:boolean, platform:string, eksik:string[],
 *           eksikSecmeli:string[], teknik:string, uyari:string, kaynak:string}}
 */
export function ortamiSinifla({
  platformOS = '',
  modul = null,
  yuklemeHatasi = null,
  kaynak = '',
} = {}) {
  const platform = String(platformOS || '').toLowerCase();
  const teknikHata = hataMetni(yuklemeHatasi);

  const temel = {
    durum: DURUM.MODUL_YOK,
    destek: false,
    platform,
    eksik: [],
    eksikSecmeli: [],
    teknik: '',
    uyari: '',
    kaynak: kaynak || '',
  };

  // 1) Tarayıcı GERÇEKTEN tarayıcıdır — burada yerel bildirim yoktur.
  //    Bu dalın tek koşulu platformun web olması; modül durumu ölçülmez.
  if (platform === 'web') {
    return { ...temel, durum: DURUM.WEB, destek: false, kaynak: '' };
  }

  // 2) Elde ÇALIŞAN bir modül varsa, önceki yükleme hatası artık ölümcül
  //    değildir (ör. barrel patladı, parçalı yükleme başardı).
  if (modul) {
    const eksik = eksikOlanlar(modul, GEREKLI_API);
    const eksikSecmeli = eksikOlanlar(modul, SECMELI_API);
    if (eksik.length) {
      return {
        ...temel,
        durum: DURUM.API_EKSIK,
        destek: false,
        eksik,
        eksikSecmeli,
        teknik: `Eksik işlev: ${eksik.join(', ')}`,
      };
    }
    return {
      ...temel,
      durum: DURUM.HAZIR,
      destek: true,
      eksikSecmeli,
      // Hata metni kaybolmaz: tanı için taşınır ama kullanıcıyı engellemez.
      uyari: teknikHata,
    };
  }

  // 3) Modül yok ama GERÇEK bir hata var → sessizce "tarayıcı" deme.
  if (teknikHata) {
    return { ...temel, durum: DURUM.MODUL_HATA, destek: false, teknik: teknikHata, kaynak: '' };
  }

  // 4) Ne modül ne hata → paket hiç kurulmamış.
  return { ...temel, durum: DURUM.MODUL_YOK, destek: false, kaynak: '' };
}

/** Ekranda gösterilecek DÜRÜST açıklama — durum ne ise o yazılır. */
export function ortamAciklamasi({ durum = '', platform = '', teknik = '' } = {}) {
  switch (durum) {
    case DURUM.WEB:
      return 'Tarayıcıda telefon bildirimi kurulamaz. Bu özellik yalnız Android/iOS '
        + 'uygulamasında çalışır; buradaki bildirim listesi çalışmaya devam eder.';
    case DURUM.HAZIR:
      return '';
    case DURUM.MODUL_YOK:
      return 'Bildirim modülü bu derlemede bulunamadı, bu yüzden telefon hatırlatması '
        + 'kurulamıyor. Uygulamanın güncel sürümünü açtığında bu bölüm kendiliğinden '
        + 'çalışır hâle gelir.';
    case DURUM.MODUL_HATA:
      return 'Bildirim modülü bu cihazda yüklenemedi, bu yüzden telefon hatırlatması '
        + `kurulamıyor (platform: ${platform || 'bilinmiyor'}). Bu bir tarayıcı sınırı değil; `
        + `cihazdaki teknik durum: ${teknik || 'ayrıntı alınamadı'}`;
    case DURUM.API_EKSIK:
      return 'Bildirim modülü yüklendi ancak yerel bildirim işlevleri eksik olduğu için '
        + `hatırlatma kurulamıyor. Teknik durum: ${teknik || 'ayrıntı alınamadı'}`;
    default:
      return 'Bildirim ortamı okunamadı; hatırlatma kurulup kurulamayacağı doğrulanamıyor.';
  }
}

/** Kısa tanı satırı (tanı amaçlı; kişisel veri içermez). */
export function ortamOzeti(o = {}) {
  const p = o.platform || 'bilinmiyor';
  const k = o.kaynak ? ` · kaynak: ${o.kaynak}` : '';
  const t = o.teknik ? ` · ${o.teknik}` : '';
  const u = !o.teknik && o.uyari ? ` · uyarı: ${o.uyari}` : '';
  return `durum: ${o.durum || 'bilinmiyor'} · platform: ${p}${k}${t}${u}`;
}

File: flutter/lib/core/push_env.dart
// KAYNAK: app/src/pushEnv.js — çeviri.
//
// TELEFON BİLDİRİMİ — ORTAM VE YETENEK SINIFLANDIRMASI (SAF MODÜL)
//
// KULLANICIYA GÖRÜNEN METİNLER VE DURUM ADLARI BİREBİR KORUNDU.
//
// ────────────────────────────────────────────────────────────────────────────
// ÇEVİRİDE DARALAN TEK PARÇA — VE NEDENİ
//
// Kaynaktaki `ortamiSinifla`, `expo-notifications` paketinin giriş dosyası
// UZAK-PUSH alt modüllerini de yüklediği ve Expo Go'da bunlar kayıtlı olmadığı
// için import'un PATLAMASINDAN doğmuştu. Sonuç: gerçek Android telefon "web"
// gibi görünüyor ve ekranda "Tarayıcıda telefon bildirimi kurulamaz." yazıyordu
// — YANLIŞ TEŞHİS. Kaynak bu yüzden modülü iki aşamalı yükleyip hangi tek tek
// işlevin elde olduğunu ölçüyordu (GEREKLI_API / SECMELI_API listeleri).
//
// Flutter'da bu sorunun karşılığı YOKTUR: `flutter_local_notifications` tek
// parça bir eklentidir, uzak-push alt modülü içermez ve derleme zamanında
// bağlanır — kısmen yüklenmesi mümkün değildir. "Hangi işlev elde?" sorusunu
// çalışma zamanında sormak bu yüzden anlamsız olurdu.
//
// KORUNAN: DURUM sözlüğü, `ortamAciklamasi` metinleri ve `ortamOzeti` biçimi
// aynen duruyor — ekran hiçbir farkla karşılaşmaz. Sınıflandırma artık iki
// GERÇEK girdiye bakar: platform ve eklentinin ilklendirilip ilklendirilemediği.

/// Ortamın ayrıştırılmış hâlleri.
abstract final class PushDurum {
  static const String web = 'web'; // gerçekten tarayıcı → yerel bildirim yok
  static const String hazir = 'hazir'; // yerel bildirim API'si elde
  static const String modulYok = 'modul-yok'; // paket bulunamadı
  static const String modulHata = 'modul-hata'; // paket var ama hata verdi
  static const String apiEksik = 'api-eksik'; // yüklendi ama işlevler yok
}

typedef PushOrtam = ({
  String durum,
  bool destek,
  String platform,
  String teknik,
  String uyari,
  String kaynak,
});

String _hataMetni(Object? e) {
  if (e == null) return '';
  return '$e'.trim();
}

/// Ortamı sınıflandırır.
///
/// [ilklendi] eklenti başarıyla kurulduysa true.
/// [yuklemeHatasi] kurulamadıysa GERÇEK hata (sessizce yutulmaz).
PushOrtam ortamiSinifla({
  String platformOS = '',
  bool ilklendi = false,
  Object? yuklemeHatasi,
}) {
  final platform = platformOS.toLowerCase();
  final teknikHata = _hataMetni(yuklemeHatasi);

  // 1) Tarayıcı GERÇEKTEN tarayıcıdır — burada yerel bildirim yoktur.
  //    Bu dalın tek koşulu platformun web olması; eklenti durumu ölçülmez.
  if (platform == 'web') {
    return (
      durum: PushDurum.web,
      destek: false,
      platform: platform,
      teknik: '',
      uyari: '',
      kaynak: '',
    );
  }

  // 2) Eklenti çalışıyorsa hazırız. Önceki bir uyarı ölümcül değildir; tanı
  //    için taşınır ama kullanıcıyı ENGELLEMEZ.
  if (ilklendi) {
    return (
      durum: PushDurum.hazir,
      destek: true,
      platform: platform,
      teknik: '',
      uyari: teknikHata,
      kaynak: 'paket',
    );
  }

  // 3) Eklenti yok ama GERÇEK bir hata var → sessizce "tarayıcı" deme.
  if (teknikHata.isNotEmpty) {
    return (
      durum: PushDurum.modulHata,
      destek: false,
      platform: platform,
      teknik: teknikHata,
      uyari: '',
      kaynak: '',
    );
  }

  // 4) Ne eklenti ne hata → henüz ilklendirilmemiş.
  return (
    durum: PushDurum.modulYok,
    destek: false,
    platform: platform,
    teknik: '',
    uyari: '',
    kaynak: '',
  );
}

/// Ekranda gösterilecek DÜRÜST açıklama — durum ne ise o yazılır.
String ortamAciklamasi({
  String durum = '',
  String platform = '',
  String teknik = '',
}) => switch (durum) {
  PushDurum.web =>
    'Tarayıcıda telefon bildirimi kurulamaz. Bu özellik yalnız Android/iOS '
        'uygulamasında çalışır; buradaki bildirim listesi çalışmaya devam eder.',
  PushDurum.hazir => '',
  PushDurum.modulYok =>
    'Bildirim modülü bu derlemede bulunamadı, bu yüzden telefon hatırlatması '
        'kurulamıyor. Uygulamanın güncel sürümünü açtığında bu bölüm kendiliğinden '
        'çalışır hâle gelir.',
  PushDurum.modulHata =>
    'Bildirim modülü bu cihazda yüklenemedi, bu yüzden telefon hatırlatması '
        'kurulamıyor (platform: ${platform.isEmpty ? 'bilinmiyor' : platform}). '
        'Bu bir tarayıcı sınırı değil; cihazdaki teknik durum: '
        '${teknik.isEmpty ? 'ayrıntı alınamadı' : teknik}',
  PushDurum.apiEksik =>
    'Bildirim modülü yüklendi ancak yerel bildirim işlevleri eksik olduğu için '
        'hatırlatma kurulamıyor. Teknik durum: '
        '${teknik.isEmpty ? 'ayrıntı alınamadı' : teknik}',
  _ =>
    'Bildirim ortamı okunamadı; hatırlatma kurulup kurulamayacağı doğrulanamıyor.',
};

/// Kısa tanı satırı (tanı amaçlı; kişisel veri içermez).
String ortamOzeti(PushOrtam? o) {
  if (o == null) return 'durum: bilinmiyor · platform: bilinmiyor';
  final p = o.platform.isEmpty ? 'bilinmiyor' : o.platform;
  final k = o.kaynak.isNotEmpty ? ' · kaynak: ${o.kaynak}' : '';
  final t = o.teknik.isNotEmpty ? ' · ${o.teknik}' : '';
  final u = o.teknik.isEmpty && o.uyari.isNotEmpty
      ? ' · uyarı: ${o.uyari}'
      : '';
  final d = o.durum.isEmpty ? 'bilinmiyor' : o.durum;
  return 'durum: $d · platform: $p$k$t$u';
}
